### DISCUSSION TITLE
Fan-out with independent Volume for each branch

### DISCUSSION THREAD

**Discussion by eduardohenriquearnold** [argoproj-labs/hera#539] [Q&A] [answered]
Hi, I would like to create a workflow with a fan-out where subsequent tasks share a volume (unique to each branch). 

The DAG would look something like:
```
             Task A
               │
   ┌───────────┼───────────┐
   │           │           │
   ▼           ▼           ▼
Task B1     Task B2     Task B3
   │           │           │
   ▼           ▼           ▼
Task C1     Task C2     Task C3
```

For example, Tasks B1, C1 share Volume 1; Tasks B2, C2 share Volume 2, etc.

I know I can define a volume to be used by two subsequent tasks as follows:
```python
from hera import Task, Workflow, Volume,
with Workflow("volumes-pvc-hera-", generate_name=True) as w:
    volumes = [Volume(size="1Gi", mount_path="/mnt/vol")]
    generate = Task(
        "generate",
        image="docker/whalesay:latest",
        command=["sh", "-c"],
        args=[
            "echo generating message in volume; cowsay hello world | tee /mnt/vol/hello_world.txt"
        ],
        volumes=volumes,
    )
​
    print = Task(
        "print",
        image="alpine:latest",
        command=["sh", "-c"],
        args=[
            "echo getting message from volume; find /mnt/vol; cat /mnt/vol/hello_world.txt"
        ],
        volumes=volumes,
    )
    generate >> print
```

As noted above, one needs the reference to the volume for each subsequent task. My question is: how to generalise the example above for the fan-out case? Thanks!

**Comment by eduardohenriquearnold** [comment]
I have also a minimal example below:

```python
def A():
    import json
    import sys

    # this can be anything! e.g fetch from some API, then in parallel process all entities; chunk database records
    # and process them in parallel, etc.
    json.dump([i for i in range(10)], sys.stdout)


def B(value: int):
    print(f"Received value: {value}!")
    with open(f"/mnt/vol/{value}.txt", 'w') as f:
        f.write("Values!")

def C(value: int):
    import pathlib
    # do something with value and files in /mnt/vol

    print("Files in /mnt/vol:")
    print(tuple(pathlib.Path("/mnt/vol").iterdir()))


# assumes you used `hera.set_global_token` and `hera.set_global_host` so that the workflow can be submitted
with Workflow("dynamic-fanout-", generate_name=True) as w:
    volumes = [Volume(size="1Gi", mount_path="/mnt/vol")]
    A_task = Task("a", A)
    B_task = Task("b", B, with_param=A_task.get_result(), volumes=volumes)
    C_task = Task("c", C, with_param=A_task.get_result(), volumes=volumes)

    A_task >> B_task >> C_task
```

When I execute this, the shared volume is the same across the branches, and Task C only starts once all Task B's have completed. My questions are:

1. How to make Task C start immediately once it's corresponding B task completes (and not wait for all B tasks to complete).
2. How to make the volume independent for different branches of tasks B, C. So each Task C should only see a single file in `/mnt/vol`

**Comment by eduardohenriquearnold** [comment]
I think I was able to solve question 1 using a parallel DAG as follows
```python
def A():
    import json
    import sys

    # this can be anything! e.g fetch from some API, then in parallel process all entities; chunk database records
    # and process them in parallel, etc.
    json.dump([i for i in range(10)], sys.stdout)


def B(value: int):
    import pathlib
    pathlib.Path("/mnt/vol").mkdir(parents=True, exist_ok=True)
    print(f"Received value: {value}!")
    with open(f"/mnt/vol/{value}.txt", "w") as f:
        f.write("Values!")


def C(value: int):
    import pathlib

    # do something with value and files in /mnt/vol

    print("Files in /mnt/vol:")
    print(tuple(pathlib.Path("/mnt/vol").iterdir()))


# assumes you used `hera.set_global_token` and `hera.set_global_host` so that the workflow can be submitted
with Workflow("dynamic-fanout-", generate_name=True, service=service, service_account_name=sa) as w:
    with DAG("pipeline", inputs=[Parameter("value")], outputs=[Parameter("value")]) as pipeline:
        volumes = [Volume(size="1Gi", mount_path="/mnt/vol")]
        B_task = Task("b", B, inputs=[pipeline.get_parameter("value")], volumes=volumes)
        C_task = Task("c", C, inputs=[pipeline.get_parameter("value")], volumes=volumes)
        B_task >> C_task
        pipeline.outputs = [C_task.get_result_as("output")]

    A_task = Task("a", A)
    parallel_task = Task("parallel-pipeline", dag=pipeline, with_param=A_task.get_result())
    A_task >> parallel_task
```

Still, the same volume is shared across all branches, so question (2) remains.

**Comment by flaviuvadan** [comment, marked as answer]
Hi @eduardohenriquearnold! Thanks for the question :) I think this is actually a limitation of Argo Workflows. When volumes are created dynamically they are set at the workflow level. So, Argo tells K8S about the PV/PVCs it needs to create. Those PV/PVCs, depending on your cluster setup, either instantly create the volumes when the workflow is created or when the pod is created/ready to mount the volume. To my knowledge, there's no way for Argo to know how many tasks there will be in the fanout, so it does not know what volumes to create. Now, I think there might be ways around this. I have not experimented with this so take it with a grain of salt. This idea is also conditioned on your application knowing how many tasks there will be a fanout over:
- take the N volumes required based on the AOT known fanout elements and create _that_ many volumes in the workflow
- set a K8S storage class on those volumes that says "only create those volumes when a pod is ready to mount them". We use this at Dyno Therapeutics by having a custom storage class in K8S with `VolumeBindingMode`set to `WaitForFirstConsumer` and that storage class is set on volumes by default
- pass the volume IDs as a parameter to the main task that fans out
- perform a zip between the fanout elements and the volume IDs
- make the fanout task mount the volume identified by ID `parameters.volume.volume_id` or something like that

I _think_ the above _might_ work? I don't think this is fully possible in Hera at the moment because we had an accidental regression and we cannot set volumes at the workflow level anymore. However, it will for sure be possible in the upcoming Hera V5, where you will get full feature parity with Argo Workflows + Events!

**Reply by flaviuvadan** [reply]
This [example](https://github.com/argoproj-labs/hera/blob/main/examples/workflows/volume_mounts.py) might help!

**Reply by eduardohenriquearnold** [reply]
That's very helpful, @flaviuvadan . Thanks for including adding these to the official examples!

The example provided assumes a known number of "fan-out" tasks or volumes to be created. It seems possible to use create fully dynamic resources, e.g. when you don't know a priori how many volumes you need to create, using [Argo Kubernetes Resource Management](https://argoproj.github.io/argo-workflows/walk-through/kubernetes-resources/). Do you know if Hera v5 supports using this feature as well? Of course, you're then responsible for cleaning up those resources afterwards.

Specifically, do you know if it's possible to convert the following Argo YAML to a Hera equivalent?
```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: generate-volume-template
spec:
  entrypoint: generate-volume
  templates:
  - name: generate-volume
    inputs:
      parameters:
      - name: pvc-size
    resource:
      action: create
      setOwnerReference: true
      manifest: |
        apiVersion: v1
        kind: PersistentVolumeClaim
        metadata:
          generateName: sequence-storage-pvc-
        spec:
          accessModes: ['ReadWriteOnce']
          resources:
            requests:
              storage: '{{inputs.parameters.pvc-size}}'
          storageClassName: zonalpd-storageclass
    outputs:
      parameters:
      - name: pvc-name
        valueFrom:
          jsonPath: '{.metadata.name}'
```

**Reply by flaviuvadan** [reply]
Yes! You can use the Hera `Resource` [object](https://github.com/argoproj-labs/hera/blob/main/src/hera/workflows/resource.py#L14) for exactly this purpose! [Here's an example](https://github.com/argoproj-labs/hera/blob/814e39cf3f0b59cce4e24a8696d836f1ff2397ff/examples/workflows/resource_flags.py#L2)

**Reply by Mastergalen** [reply]
Indeed, you can create PVCs dynamically:

```python
generate_pvc = Resource(
        name="generate-pvc",
        inputs=Parameter(name="pvc-size"),
        action="create",
        set_owner_reference=True,
        outputs=[
            Parameter(
                name="pvc-name", value_from=m.ValueFrom(json_path="{.metadata.name}")
            )
        ],
        manifest="""
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  generateName: hera-scratch-pvc-
spec:
  accessModes: ['ReadWriteOnce']
  resources:
    requests:
      storage: '{{inputs.parameters.pvc-size}}'
  storageClassName: standard-rwo
            """,
    )
```

But... Argo will not delete the PVC unless you delete the workflow itself. You therefore need a final step that cleans up your PVC:

```python
delete_pvc = Resource(
        name="delete-pvc",
        inputs=Parameter(name="pvc-name"),
        action="delete",
        flags=[
            "pvc",
            "{{inputs.parameters.pvc-name}}",
            "--grace-period=0",
            "--force",
        ],
    )

whalesay = Container(
    name="whalesay",
    image="docker/whalesay:latest",
    command=["sh", "-c"],
    inputs=[
        Parameter(name="pvc-name"),
    ],
    args=[
        "echo generating message in volume {{inputs.parameters.pvc-name}}; cowsay hello world | tee /mnt/vol/hello_world.txt"
    ],
    volume_mounts=[m.VolumeMount(name="workdir", mount_path="/mnt/vol")],
    volumes=m.Volume(
        name="workdir",
        persistent_volume_claim={
            "claim_name": "{{inputs.parameters.pvc-name}}"
        },
    ),
)

with Steps(name=name, inputs=inputs) as s:
    Step(
        name="generate-volume",
        template=generate_pvc.name,
        arguments={"pvc-size": pvc_size},
    )

    whalesay(
        name="whalesay",
        arguments={
            "pvc-name": "{{steps.generate-volume.outputs.parameters.pvc-name}}",
        },
    )

    # Ensure PVC is deleted when context is exited
    Step(
        name="delete-pvc",
        template=delete_pvc.name,
        arguments={
            "pvc-name": "{{steps.generate-volume.outputs.parameters.pvc-name}}"
        },
    )
```

Furthermore, you need to set `podGC` to `OnPodCompletion`, so that pods get deleted immediately and release the finalizer set on the PVC. Otherwise your delete-volume will get stuck in a `Pending` state.

Another caveat: If your `whalesay` step fails for whatever reason, `delete-pvc` will not be run and your PVC will still persist until the Workflow is deleted. To clean that up you could try a lifecycle hook (which I couldn't get work yet) or set a [workflow TTL](https://github.com/argoproj/argo-workflows/blob/master/examples/gc-ttl.yaml)

**Reply by flaviuvadan** [reply]
Thanks for outlining the above @Mastergalen! I think another option for cleaning up is to use exit handlers, either at the task/step level or the workflow level. I think these will run irrespective of result! Can also run as a normal step but then make the step "continue on fail"

**Reply by Mastergalen** [reply]
@flaviuvadan Indeed, that is what I have been trying with lifecycle hooks, but ran into issues with the `workflow.status` not being accessible in template-level hooks haha 😅 .

See https://github.com/argoproj/argo-workflows/issues/11269

**Reply by flaviuvadan** [reply]
That's annoying :( thank you for explaining!

**Reply by Muralikrishna2009** [reply]
@Mastergalen In the below - The volume doesnot work in Container(). Can you please verify it once. May be a bug?

volumes=m.Volume(
        name="workdir",
        persistent_volume_claim={
            "claim_name": "{{inputs.parameters.pvc-name}}"
        },
```
whalesay = Container(
    name="whalesay",
    image="docker/whalesay:latest",
    command=["sh", "-c"],
    inputs=[
        Parameter(name="pvc-name"),
    ],
    args=[
        "echo generating message in volume {{inputs.parameters.pvc-name}}; cowsay hello world | tee /mnt/vol/hello_world.txt"
    ],
    volume_mounts=[m.VolumeMount(name="workdir", mount_path="/mnt/vol")],
    volumes=m.Volume(
        name="workdir",
        persistent_volume_claim={
            "claim_name": "{{inputs.parameters.pvc-name}}"
        },
    ),
)
```

**Reply by Muralikrishna2009** [reply]
@eduardohenriquearnold - Can you please share how you made this work, will be very useful for my fanout use case
An example will help
